### ISSUE TITLE
[question] Need clarification on the purpose and performance benefits of GarbageCollection class

### ISSUE THREAD

**Issue by qsh-zh** [pytorch/torchtitan#696]
For the [impl](https://github.com/pytorch/torchtitan/blob/5525d7723175a1b4477bde3034a96f803b6c3fae/torchtitan/utils.py#L104)

I have several questions about the motivation and use cases for this class:

Could you provide examples of scenarios where this class can improves performance? compare against default Python GC?

To my understanding, during backward, activation cuda memory should be released timely when we run backward in computational graph, will the GarbageCollection affect how we release cuda memory?

What are the tradeoffs of disabling automatic GC (gc.disable())?

**Comment by tianyu-l** [comment]
This was added in https://github.com/pytorch/torchtitan/pull/148. You can check the discussions there for more context.

cc: @lessw2020 @yifuwang -- The GC change we made seems worth better documentation.

**Comment by fegin** [comment]
In distributed training, Python's garbage collection (GC) can kick in at different times for different ranks, leading to slow communication and stragglers. To avoid this issue, we need to perform manual GC to ensure all ranks collect garbage simultaneously. Consider a scenario where we're using FSDP and Python triggers garbage collection for rank0 during the first allgather, rank1 during the second allgather, and so on. In this case, communication will be severely slowed down due to the straggling effect caused by asynchronous garbage collection.

While disabling Python GC might seem like an easy solution, it's not feasible due to potential circular dependencies in the code. We still rely on GC to recycle some Python objects. Disabling GC altogether could lead to CPU or GPU memory exhaustion.

So, performing manual GC is the best approach to resolve this issue.

**Comment by qsh-zh** [comment]
hi @fegin  thank for your response. If we disable GC and run collection manually, will it fail to release CUDA memory in a timely manner? For example, during the backward pass, activations from later layers should ideally be released after backpropagation uses them for gradient computation in the previous layer as GC not collect zero reference count variables?

**Comment by qsh-zh** [comment]
@yifuwang @fegin  thank you for explanation! looks good to me
